Support for the Docker Python SDK

apierleoni opened the pull request:
## Description
Hi there, we did some work to run a data processing pipeline using Docker containers with Luigi using the Docker Python SDK and we would like to contribute it back.

## Motivation and Context
These PR allows to use the Docker Python SDK to run Docker containers instead of launching them from a subprocess.
The main advantage over calling an external process is that all the communications done with the Docker daemon are through the API and not a subprocess, being able to handle exceptions properly and potentially connect to remote APIs (we did not test this scenario)
It also solve non trivial problems with the docker SDK like mounting a single file, and makes sure that there is enough space in the container to run the job by mounting a temporary directory from the host , avoiding the container size limit of 10Gb.

## Have you tested this? If so, how?
Unit tests are included. But they require a Docker daemon to run. Not sure it is available in Travis.
We are currently running this in our staging environment and we are ready to run production pipelines with this contrib.

elipapa commented:
@Tarrasch two of the travis tests are timing out, but it seems something common to all previous PRs

Tarrasch commented:
@elipapa, yes sorry it isn't fixed yet. 

So far this seems promising. Can you please contact the authors of https://github.com/spiside/docker-luigi and see how your efforts can be joined (or how they differ)? Most ideally code lives in a separate repo (to minimize burden to maintain the main repo).

elipapa commented:
Thanks for the pointer to the `docker-luigi` project. We had looked at it,
but it solves more of a deployment problem (ie. how to run luigi *in* a
container environment).

Our contribution allows to run _containerized tasks_ and thus it's a
contribution much closer in spirit and place to the other ones in [contrib](
https://github.com/spotify/luigi/tree/master/luigi/contrib), such as
`ExternalTask`, `HiveTask`, `KubernetesJobTask` and others..
Which is why I really don't see how it could live in a separate repo.

Just for the sake of comparison, airflow has had [a Docker Operator](
https://airflow.incubator.apache.org/_modules/airflow/operators/docker_operator.html)
since the beginning.

On Fri, Jun 23, 2017 at 10:33 AM, Arash Rouhani <notifications@github.com>
wrote:

> @elipapa <https://github.com/elipapa>, yes sorry it isn't fixed yet.
>
> So far this seems promising. Can you please contact the authors of
> https://github.com/spiside/docker-luigi and see how your efforts can be
> joined (or how they differ)? Most ideally code lives in a separate repo (to
> minimize burden to maintain the main repo).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spotify/luigi/pull/2158#issuecomment-310619916>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAG34oNmOoF_S7OzWp9MWWWCpU-IZXJgks5sG4ZSgaJpZM4N7Z66>
> .
>

Tarrasch commented:
Okay cool! Is this ready to be reviewed or do you still append edits to this PR? I can do a quick review to see if it seems okay.

> Our contribution allows to run _containerized tasks_ and thus it's a contribution much closer in spirit and place to the other ones in [contrib]( https://github.com/spotify/luigi/tree/master/luigi/contrib), such as `ExternalTask`, `HiveTask`, `KubernetesJobTask` and others..

You can part of this in the module level documentation by the way (if not there already). This is great information. :)

elipapa commented:
Ready to be reviewed afaik.
Let us know if you think it needs more module docs.



________________________________
From: Arash Rouhani <notifications@github.com>
Sent: Monday, June 26, 2017 9:33:06 PM
To: spotify/luigi
Cc: elipapa; Mention
Subject: Re: [spotify/luigi] Support for the Docker Python SDK (#2158)


Okay cool! Is this ready to be reviewed or do you still append edits to this PR? I can do a quick review to see if it seems okay.

Our contribution allows to run containerized tasks and thus it's a contribution much closer in spirit and place to the other ones in contrib<https://github.com/spotify/luigi/tree/master/luigi/contrib>, such as ExternalTask, HiveTask, KubernetesJobTask and others..

You can part of this in the module level documentation by the way (if not there already). This is great information. :)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/spotify/luigi/pull/2158#issuecomment-311173564>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAG34ltvMKufx3KvBF0V7yKgJzUy-Wpqks5sIBWCgaJpZM4N7Z66>.

apierleoni commented:
Thanks @Tarrasch, I'll extend the docs as well, did not think of that before...

dlstadther commented:
Unrelated to the Docker review, it would be nice if your addition to "Companies who use Luigi" were its own PR. We could go ahead and merge that part.

tomplex commented:
Any word on this PR? I would love to use this functionality.

dlstadther commented:
I think we're waiting on doc extension and the separation of "Companies who use Luigi", @tomplex .

Any update @apierleoni

apierleoni commented:
Hi, sorry I have been off for a while.
@Tarrasch 
I have extended the module docs to include the DockerTask use case. Not sure if you prefer this docs to be somewhere else. If yes, can you point me to it?
@dlstadther 
Happy to do an other PR for the link to our blog if this one requires reviews.

dlstadther commented:
@apierleoni It would be great if you could separate "companies who use luigi" into another PR - that one can be merged quickly.

Also, looks like you've had some Travis errors; could you take a look and resolve? Thanks!

elipapa commented:
@dlstadther the travis errors are not ours and apply to every PR. @Tarrasch has already acknowledged that they have yet to be fixed (see comment above: june 23)

on separating the PR: we have no problem waiting for this one to be merged for our "companies with luigi" link to show up. We'd rather keep them in the same PR for simplicity and just wait for the review for it to be merged.

dlstadther commented:
@elipapa Those errors were resolved elsewhere. Your current errors are flake8.

apierleoni commented:
flake8 errors fixed now, all tests passed on travis

dlstadther commented on tox.ini:
```diff
@@ -10,6 +10,7 @@ deps=
   moto<1.0
   HTTPretty==0.8.10
   nose<2.0
+  docker
```
Do you want to specify a specific version?

apierleoni replied:
yes thanks, well spotted!

dlstadther commented:
If there are no objections, i'm ok to merge this

drobakowski commented on test/contrib/docker_runner_test.py:
```diff
@@ -0,0 +1,162 @@
+    image = "busybox"
+    name = "MountLocalFileAsVolume"
+    # volumes= {'/tmp/local_file_test': {'bind': local_file.name, 'mode': 'rw'}}
+    volumes = [local_file.name + ':/tmp/local_file_test']
```
Shouldn't this be renamed to `binds`?

drobakowski commented on test/contrib/docker_runner_test.py:
```diff
@@ -0,0 +1,162 @@
+    dummyopt = luigi.Parameter()
+    image = "busybox"
+    name = "MountLocalFileAsVolumeWithParam"
+    volumes = [local_file.name + ':/tmp/local_file_test']
```
Shouldn't this be renamed to `binds`?

drobakowski commented on test/contrib/docker_runner_test.py:
```diff
@@ -0,0 +1,162 @@
+    name = "MountLocalFileAsVolumeWithParamRedef"
+
+    @property
+    def volumes(self):
```
Shouldn't this be renamed to `binds`?

drobakowski commented on luigi/contrib/docker_runner.py:
```diff
@@ -0,0 +1,245 @@
+        # derive volumes (ie. list of container destination paths) from
+        # specified binds
+        self._volumes = [b.split(':')[1] for b in self._binds]
+
```
One addition I had to make to get this properly working in our scenario was to include a custom instance attribute `self._complete = False` and set this to `True` when the container exited.

drobakowski commented on luigi/contrib/docker_runner.py:
```diff
@@ -0,0 +1,245 @@
+                                                      **self.container_options)
+            self._client.start(container['Id'])
+
+            exit_status = self._client.wait(container['Id'])
```
+`self._complete = True`

I'm not sure anymore if I included this simply to `True` on purpose or if this should be set in subject to the `exit_status`.

In our use case we're chaining multiple `DockerTask` consecutive so that each task depend on the previous one and its output. As I'm quite new to working with luigi I don't have that much deep knowledge and experience so far, but I think I had some problems with the luigi scheduler while the scheduler was trying to resolve and start the `DockerTask` process chain properly. After including the previous changes and overwriting the `complete` methods, the chain worked fine for me. As far as I remember correctly, I've even got the tests to run fine after renaming `volumes` to `binds`. Perhaps this is use case specific, but maybe someone with better knowledge of luigi can overlook this better.

```python
class ImportProcess(DockerTask):
    ....
    
    def complete(self):
        return self.output().exists() or self._complete
    ...
```
Of course this could also be implemented in the `DockerTask` class directly with maybe some modification like this: 
```python
class DockerTask(luigi.Task):
    ....
    def complete(self):
        return super(DockerTask, self).complete() or self._complete
```

drobakowski commented:
I do have two more smaller additions, but in order to keep this PR simple, I will create a PR after this one will be merged. Thanks again for `DockerTask`, this helps us a lot! 👍

dlstadther commented:
@apierleoni Where does this PR stand given @drobakowski 's review and comments?

apierleoni commented:
@drobakowski 
I am glad DockerTask was useful for you, and than you for the review.
Yes I have renamed `volume` to `binds`, and changed default behaviour for tmp directory mounting. Tests are now passing in my system.
I have added docker to the services in travis, so now the test should run in CI

Regarding the second issue, I am not 100% sure I understood what you want to do with the `self._complete` attribute. The way we use DockerTask is to make sure that the Target is labeled as completed, so that Luigi will know if he needs to rerun the task or not.
Would you prefer to trigger an error in case the container quit status 0? how do you know it is not completed at that stage?

@dlstadther
tests should be improved now, maybe you can help here on what is the preferred way to store a task to be completed.

apierleoni commented:
some tests are failing for an error that looks related to boto non importing the google cloud platform module. this might not be limited to this PR
https://travis-ci.org/spotify/luigi/jobs/279538792#L565

apierleoni commented:
merged with current master, still failing for boto

apierleoni commented:
@dlstadther 
I have fixed the errors with boto in the tests and resolved conflicts with master.
Should be clear to merge if you want.

apierleoni commented:
I guess the codecov issue is related with this https://github.com/codecov/support/issues/180

dlstadther commented:
Hey @apierleoni , do you have any insight into why [two of the Docker tests](https://travis-ci.org/spotify/luigi/jobs/342494179) started failing recently?

It'd be much appreciated if you could check it out! Thanks

apierleoni commented:
@dlstadther sorry for the late reply, is it still an issue? I cannot find the failing tests.

dlstadther commented:
@apierleoni They've since been fixed!
